Kernel/Storage/SD/SDHostController.cpp

```diff
@@ -0,0 +1,667 @@
+    VERIFY(host_version() == SD::HostVersion::Version3);
+    u32 divisor = AK::max(sd_clock_frequency / (frequency), 2);
+
+    if (sd_clock_frequency / divisor >= frequency) {
+        divisor++;
+    }
+
+    divisor -= 2;
+
+    // 2. Set **Internal Clock Enable** and **SDCLK Frequency Select** in the
+    // *Clock Control* register
+    const u32 two_upper_bits_of_sdclk_frequency_select = (divisor >> 8 & 0x3)
+        << 6;
+    const u32 eight_lower_bits_of_sdclk_frequency_select = (divisor & 0xff)
+        << 8;
+    const u32 sdclk_frequency_select = two_upper_bits_of_sdclk_frequency_select | eight_lower_bits_of_sdclk_frequency_select;
+    m_registers->host_configuration_1 = m_registers->host_configuration_1 | internal_clock_enable | sdclk_frequency_select;
+
+    // 3. Check **Internal Clock Stable** in the *Clock Control* register until
+    // it is 1
+    if (!retry_with_timeout(
+            [&] {
+                return m_registers->host_configuration_1 & internal_clock_stable;
+            })) {
+        return EIO;
+    }
+
+    // 4. Set **SD Clock Enable** in the *Clock Control* register to 1
+    m_registers->host_configuration_1 = m_registers->host_configuration_1 | sd_clock_enable;
+
+    return {};
+}
+
+void SDHostController::sd_clock_stop()
+{
+    // 3.2.2 SD Clock Stop Sequence
+
+    // 1. Set **SD Clock Enable** in the *Clock Control* register to 0
+    m_registers->host_configuration_1 = m_registers->host_configuration_1 & ~sd_clock_enable;
+}
+
+ErrorOr<void> SDHostController::sd_clock_frequency_change(u64 new_frequency)
+{
+    // 3.2.3 SD Clock Frequency Change Sequence
+
+    // 1. Execute the SD Clock Stop Sequence
+    sd_clock_stop();
+
+    // 2. Execute the SD Clock Supply Sequence
+    return sd_clock_supply(new_frequency);
+}
+
+ErrorOr<void> SDHostController::reset_host_controller()
+{
+    m_registers->host_configuration_0 = 0;
+    m_registers->host_configuration_1 = m_registers->host_configuration_1 | software_reset_for_all;
+    if (!retry_with_timeout(
+            [&] {
+                return (m_registers->host_configuration_1 & software_reset_for_all) == 0;
+            })) {
+        return EIO;
+    }
+
+    return {};
+}
+
+ErrorOr<void> SDHostController::transaction_control_with_data_transfer_using_the_dat_line_without_dma(
+    SD::CommandIndex index,
+    u32 argument,
+    u32 block_count,
+    u32 block_size,
+    UserOrKernelBuffer buf,
+    DataTransferType data_transfer_type)
+{
+    VERIFY(block_size * block_count % 4 == 0);
+    auto command = SD::get_command(index);
+    // 3.7.2 Transaction Control with Data Transfer Using DAT Line (without DMA)
+
+    // 1. Set the value corresponding to the executed data byte length of one
+    // block to Block Size register.
+    // 2. Set the value corresponding to the executed data block count to Block
+    // Count register in accordance with Table 2-8.
+    m_registers->block_size_and_block_count = (block_count << 16) | block_size;
```

Comment:
why arent these 2 fields instead?

Reply:
Not sure how relevant this is (maybe the compiler already handles this) but the spec for the RPi says that ["the EMMC module registers can only be accessed as 32 bit registers"](https://pdf1.alldatasheet.com/pdfjsview/web/viewer.html?file=//pdf1.alldatasheet.com/datasheet-pdf/view/502533/BOARDCOM/BCM2835/+_035W5VvMzS..SReHIuCHEh+/datasheet.pdf) (page 66, first paragraph of "Registers").
I wanted to be explicit when writing to these registers.

Also, in case, should I still change this after what @supercomputer7 said above?

Reply:
If this is directly read from MMIO then you definitely should leave it 32 bit, the `volatile` should restrict it to only do full 32 bit reads then, IIRC